### PR TITLE
MULE-11002: add cache for XML factories to avoid configuring them (wi…

### PR DESCRIPTION
**Pull request by alepulver**
…th setFeature) which is expensive

### CODE REVIEW

File: core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.util.xmlsecurity;
+
+/**
+ * Cache key for XMLSecureFactoriesCache. Also creates the corresponding factory.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java]: link in javadoc

File: core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public abstract class XMLFactoryConfig
+{
+    protected Boolean externalEntities;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java]: are this required to be protected? cannot be private?

File: core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        final int primeNumber = 31;
+
+        int result = this.externalEntities.hashCode();
+        result = primeNumber * result + this.expandEntities.hashCode();
+        result = primeNumber * result + this.factoryName.hashCode();
+
+        return result;
+    }
+
+    public abstract Object createFactory();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java]: is there no common superclass to what is being built?

File: core/src/main/java/org/mule/util/xmlsecurity/XMLSecureFactories.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        XMLSecureFactories factory = new XMLSecureFactories();
+
+        factory.externalEntities = externalEntities;
+        factory.expandEntities = expandEntities;
+
+        return factory;
+    }
+
+    public static XMLSecureFactories createDefault()
```
Comment (elrodro83) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLSecureFactories.java]: why not getDefault and return always the same instance?

File: core/src/main/java/org/mule/util/xmlsecurity/XMLSecureFactoriesImpl.java [hunk 5]
```diff
@@ -18,40 +17,13 @@
 import org.apache.commons.logging.LogFactory;
 
 /**
- * Provide XML parser factories configured to avoid XXE and BL attacks according to global configuration (safe by default)
+ * Create different XML factories configured through the same interface for disabling vulnerabilities.
  */
-public class XMLSecureFactories
+public class XMLSecureFactoriesImpl
```
Comment (elrodro83) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLSecureFactoriesImpl.java]: rename to DefualtXMLSecureFactories

File: core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java (hunk 2 above)
Comment (alepulver) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java]: If they are private they would require getters to be accessed from the conflict classes that implement createFactory.

File: core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java (hunk 3 above)
Comment (alepulver) [inline core/src/main/java/org/mule/util/xmlsecurity/XMLFactoryConfig.java]: No, they are different interfaces to XML parsers made at different times. All "standard" but not compatible :)

**Pull request merged**